### PR TITLE
Add extension types for console actions and topology details panel

### PR DESCRIPTION
**Pull request by rohitkrai03**
**Fixes**: https://issues.redhat.com/browse/ODC-5673
**Fixes**: https://issues.redhat.com/browse/ODC-5675

**Solution Description**: Add types for new extensions around console actions and topology details panel

### CODE REVIEW

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/actions.ts [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+namespace ExtensionProperties {
+  export type ActionCreator = {
+    /** ID used to identify the action creator. */
+    id: string;
+    /** The context ID helps to narrow the scope of contributed actions to a particular area of the application. Ex - topology, helm */
+    contextId: string | 'resource';
+    /** A react hook which creates and returns action for the given scope.
+     * If contextId = `resource` then the scope will always be a K8s resource object
+     * */
+    creator?: EncodedCodeRef;
```
Comment (christianvogt) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/actions.ts:15]:
```suggestion
    creator: EncodedCodeRef;
```

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/actions.ts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  };
+
+  export type ActionProvider = {
+    /** The context ID helps to narrow the scope of contributed actions to a particular area of the application. Ex - topology, helm */
+    contextId: string | 'resource';
+    /** A react hook which returns action for the given scope.
+     * If contextId = `resource` then the scope will always be a K8s resource object
+     * */
+    provider?: EncodedCodeRef;
```
Comment (christianvogt) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/actions.ts:24]:
```suggestion
    provider: EncodedCodeRef;
```

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/actions.ts [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+  export type ResourceActionProvider = {
+    /** The model for which this provider provides actions for. */
+    model: {
+      group: string;
+      version?: string;
+      kind?: string;
+    };
+    /** The action ids to contribute to */
+    provider?: string[];
```
Comment (christianvogt) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/actions.ts:35]:
```suggestion
    actions: string[];
```

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/actions.ts [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+export type Action = {
+  /** A unique identifier for this action. */
+  id: string;
+  /** The label to display in the UI. */
+  label: string;
+  /** Executable callback or href.
+   * External links should automatically provide an external link icon on action.
+   * */
+  cta: () => void | { href: string; external: boolean };
```
Comment (christianvogt) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/actions.ts:140]:
```suggestion
  cta: () => void | { href: string; external?: boolean };
```

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/topology-details.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
+import * as React from 'react';
+import { Extension } from '@console/plugin-sdk/src/typings/base';
+import { GraphElement } from '@patternfly/react-topology';
+import { CodeRef, EncodedCodeRef, UpdateExtensionProperties } from '../types';
+
+namespace ExtensionProperties {
```
Comment (christianvogt) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/topology-details.ts:6]: Ideally these definitions are contained in the package that owns them. But for now putting them in the main SDK is fine.

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/actions.ts (hunk 1 above)
Comment (rohitkrai03) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/actions.ts]: Done.

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/actions.ts (hunk 3 above)
Comment (rohitkrai03) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/actions.ts]: Done.

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/actions.ts (hunk 2 above)
Comment (rohitkrai03) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/actions.ts]: Done.

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/actions.ts (hunk 4 above)
Comment (rohitkrai03) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/actions.ts]: Done.

File: frontend/packages/console-dynamic-plugin-sdk/src/utils/common.ts [hunk 6]
```diff
@@ -33,3 +33,9 @@ export type ExtensionK8sVerb =
   | 'delete'
   | 'deletecollection'
   | 'watch';
+
+// Type for extension hook
+export type ExtensionHook<T, R = any> = (options: R) => ExtensionHookResult<T>;
+
+// Type for extension hook result that returns [data, resolved, error]
+export type ExtensionHookResult<T> = [T, boolean, any];
```
Comment (christianvogt) [inline frontend/packages/console-dynamic-plugin-sdk/src/utils/common.ts:41]: I'm not sure if we want to dictate a standard type here. It's really use case dependent.

thoughts @vojtechszocs

Comment (rohitkrai03) [reply frontend/packages/console-dynamic-plugin-sdk/src/utils/common.ts:41]: I had same types being used in catalog extensions. That's why I pulled it into a common type that can be used in both places.

**Pull request merged**